api/types/authority.go

```diff
@@ -479,102 +426,23 @@ func (ca *CertAuthorityV2) SetSigningAlg(alg string) {
 	ca.Spec.SigningAlg = ParseSigningAlg(alg)
 }
 
-// Check checks if all passed parameters are valid
-func (ca *CertAuthorityV2) Check() error {
-	err := ca.ID().Check()
-	if err != nil {
-		return trace.Wrap(err)
-	}
-
-	switch ca.GetType() {
-	case UserCA, HostCA:
-		err = ca.checkUserOrHostCA()
-	case JWTSigner:
-		err = ca.checkJWTKeys()
-	default:
-		err = trace.BadParameter("invalid CA type %q", ca.GetType())
-	}
-	if err != nil {
-		return trace.Wrap(err)
-	}
-	return nil
-}
-
-func (ca *CertAuthorityV2) checkUserOrHostCA() error {
-	if len(ca.Spec.CheckingKeys) == 0 {
-		return trace.BadParameter("certificate authority missing SSH public keys")
-	}
-	if len(ca.Spec.TLSKeyPairs) == 0 {
-		return trace.BadParameter("certificate authority missing TLS key pairs")
-	}
-
-	_, err := ca.Checkers()
-	if err != nil {
-		return trace.Wrap(err)
-	}
-	_, err = ca.Signers()
-	if err != nil {
-		return trace.Wrap(err)
-	}
-
-	// This is to force users to migrate
-	if len(ca.Spec.Roles) != 0 && len(ca.Spec.RoleMap) != 0 {
-		return trace.BadParameter("should set either 'roles' or 'role_map', not both")
-	}
-	if err := RoleMap(ca.Spec.RoleMap).Check(); err != nil {
-		return trace.Wrap(err)
-	}
-	return nil
-}
-
-func (ca *CertAuthorityV2) checkJWTKeys() error {
-	// Check that some JWT keys have been set on the CA.
-	if len(ca.Spec.JWTKeyPairs) == 0 {
-		return trace.BadParameter("missing JWT CA")
-	}
-
-	var err error
-	var privateKey crypto.Signer
-
-	// Check that the JWT keys set are valid.
-	for _, pair := range ca.Spec.JWTKeyPairs {
-		if len(pair.PrivateKey) > 0 {
-			privateKey, err = utils.ParsePrivateKey(pair.PrivateKey)
-			if err != nil {
-				return trace.Wrap(err)
-			}
-		}
-		publicKey, err := utils.ParsePublicKey(pair.PublicKey)
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		_, err = jwt.New(&jwt.Config{
-			Algorithm:   defaults.ApplicationTokenAlgorithm,
-			ClusterName: ca.Spec.ClusterName,
-			PrivateKey:  privateKey,
-			PublicKey:   publicKey,
-		})
-		if err != nil {
-			return trace.Wrap(err)
-		}
-	}
-
-	return nil
-}
-
 // CheckAndSetDefaults checks and set default values for any missing fields.
 func (ca *CertAuthorityV2) CheckAndSetDefaults() error {
 	err := ca.Metadata.CheckAndSetDefaults()
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	err = ca.Check()
-	if err != nil {
+	if err = ca.ID().Check(); err != nil {
 		return trace.Wrap(err)
 	}
 
-	return nil
+	switch ca.GetType() {
+	case UserCA, HostCA, JWTSigner:
+		return nil
+	default:
+		return trace.BadParameter("invalid CA type %q", ca.GetType())
+	}
```

Comment:
Minor issue: I think we have a function to check if a string is in a slice in utils. Might read better here.

Reply:
Since the values are `CertAuthType` rather than strings, you have to do a lot of type casting to use `utils.SliceContainsStr`. Ends up being much uglier
```
if !utils.SliceContainsStr([]string{string(UserCA), string(HostCA), string(JWTSigner)}, string(ca.GetType())) {
	return trace.BadParameter("invalid CA type %q", ca.GetType())
}
return nil
```
I think the switch is fine here.